### PR TITLE
front: output table does not display schedule for op via

### LINKED ISSUE #8453: Imported trains requested times are not visible in the output table
### What happened?

I imported trains from open data between RES and QR, and none of them have the information of the requested times, but most have the yellow clock icon idincating that some times have not been matched:

image

Even though they clearly try to respect specific times if we look at the space-speed-chart:

image

### What did you expect to happen?

I'm expecting the requested times to be visible in the output table, in order to understand what hour has not been respected.

### How can we reproduce it (as minimally and precisely as possible)?

1. Import trains on France between RES and QR

### On which environments the bug occurs?

Recette (SNCF)

### On which browser the bug occurs?

Google Chrome

### OSRD version (top right corner `Account` button > `Informations`)

ab7706e

### PR DESCRIPTION
**Pull request by clarani**
closes #8453

Because of the condition `if (pathStepKey in pathStepsByUic && nextOpPoint)` (l95 in `useOutputData`), the schedules of the pathSteps whose location was not a uic were ignored.

Now, we handle the pathSteps first, then the suggestedOp, and finally we merge them using the positionOnPath.

### CODE REVIEW

**Comment by bloussou** [comment]
Tested ✅

**Review by bloussou — APPROVED** [review]
Tested ✅

**Comment by emersion** [comment]
Needs to be rebased on top of https://github.com/OpenRailAssociation/osrd/commit/f0a220a785869208208e26324110725ea9aba048 (sorry!).

**Review by Akctarus — APPROVED** [review]
Tested and LGTM

File: front/src/modules/timesStops/hooks/useOutputTableData.ts [hunk 1]
```diff
@@ -65,51 +73,61 @@ function useOutputTableData(
     (opPoint) => `${opPoint.name}-${opPoint.ch}`
```
Comment (emersion) [inline front/src/modules/timesStops/hooks/useOutputTableData.ts:73]: Unrelated drive-by comment: we probably shouldn't use "name" as a key, it's not guaranteed to be unique.

File: front/src/modules/timesStops/hooks/useOutputTableData.ts [hunk 2]
```diff
@@ -65,51 +73,61 @@ function useOutputTableData(
...
-        return { ...sugOpPoint, calculatedArrival: secToHoursString(opPoint.time, true) };
-      }),
-    [simulatedTrain, pathProperties, operationalPoints, selectedTrainSchedule, pathLength]
-  );
+      return { ...sugOpPoint, calculatedArrival: secToHoursString(opPoint.time, true) };
+    });
+
+    return suggestedOpRows.map((sugOpPoint) => {
+      const matchingPathStep = pathStepRows.find(
+        (pathStepRow) => sugOpPoint.positionOnPath === pathStepRow.positionOnPath
```
Comment (emersion) [inline front/src/modules/timesStops/hooks/useOutputTableData.ts:120]: I must admit I'm not a fan of using the position on path as a kind of unique identifier… The backend doesn't guarantee that the `positionOnPath` are unique for each OP. Ideally we should also match the trigram/UIC/etc with e.g. `matchPathStepAndOp()` when we're hitting duplicate `positionOnPath` values. But it sounds unlikely to happen with the France infrastructure, so maybe we can just leave this as a TODO or something…

Edit: turns out it does happen quite often in practice.

File: front/src/modules/timesStops/hooks/useOutputTableData.ts [hunk 3]
```diff
@@ -65,51 +73,61 @@ function useOutputTableData(
...
-            ...formattedScheduleData,
-            onStopSignal: schedule?.on_stop_signal || '',
-            calculatedArrival,
-            calculatedDeparture:
-              opPoint.duration > 0 ? secToHoursString(opPoint.time + opPoint.duration, true) : '',
-            ...marginsData,
-          } as SuggestedOP;
-        }
+      if (pathStepKey in pathStepsByUic) {
+        const pathStepId = pathStepsByUic[pathStepKey].id || '';
```
Comment (emersion) [inline front/src/modules/timesStops/hooks/useOutputTableData.ts:96]: Any reason why we're still using the uic+ch here to fish the `pathStepId` (and then the `schedule`) instead of using a logic similar to the one below?

File: front/src/modules/timesStops/hooks/useOutputTableData.ts (hunk 2 above)
Comment (clarani) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts:120]: Yes, I agree :/ I have a refacto in mind, can we leave it like this for now ?

File: front/src/modules/timesStops/hooks/useOutputTableData.ts (hunk 3 above)
Comment (clarani) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts:96]: I didn't change the logic here. I can leave it like this for now if it's ok for you ?

File: front/src/modules/timesStops/hooks/useOutputTableData.ts (hunk 2 above)
Comment (emersion) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts:120]: Sure

File: front/src/modules/timesStops/hooks/useOutputTableData.ts (hunk 3 above)
Comment (emersion) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts:96]: Sure

**Review by emersion — APPROVED** [review]
Still quite fragile, but it's an improvement.

File: front/src/modules/timesStops/hooks/useOutputTableData.ts [hunk 4]
```diff
@@ -65,51 +73,61 @@ function useOutputTableData(
...
-          operationalPoints,
-          pathStepsWithOpPointIndices,
-          sugOpIndex
-        );
-        const pathStepKey = `${sugOpPoint.uic}-${sugOpPoint.ch}`;
+  const outputTableData = useMemo(() => {
+    const pathStepRows = pathStepsWithPositionOnPath.map((pathStep) => {
+      const schedule = scheduleByAt[pathStep.id];
+      const scheduleData = computeScheduleData(schedule, selectedTrainSchedule.start_time);
+      return { ...pathStep, ...formatScheduleData(scheduleData) };
```
Comment (SharglutDev) [inline front/src/modules/timesStops/hooks/useOutputTableData.ts:80]: We use this combination two times in the file (computeScheduleData then format the result). Maybe we could do the format inside the compute as format is never used without computing first ?

**Review by SharglutDev — APPROVED** [review]
Lgtm (just one comment that you can resolve if you don't agree)

File: front/src/modules/timesStops/hooks/useOutputTableData.ts (hunk 4 above)
Comment (clarani) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts:80]: Yes I agree, but I would like to do it separately if it's ok for you :)
